### ISSUE TITLE
Dynamic DNS (dyndns) - FreeDNS Update broken

### ISSUE THREAD

**Issue by sraspe** [opnsense/plugins#2428]
**Describe the bug**
The FreeDNS service will not update the IP anymore due to a broken URL. Teh first failure in my logs was on May 27 2021. The log shows a 404 error.

**To Reproduce**
1. Install the os-dyndns package
2. Create a Dynamic DNS configuration in "Services->Dynamic DNS"
3. Set Service type as "freeDNS"
4. Set Interface to monotor as your WAN
5. Set hostname to your freeDNS hostname
6. Leave Username blank
7. Set the freeDNS TOKEN as your password (can be found on the freeDNS website as your direct link or in the CURL script it generates for you)
8. Save and Force update

**Expected behavior**
The Dashboard app and FreeDNS site show your WAN IP as cached or as the A record.

**Relevant log files**
From the /var/cache/dyndns_wan_MYDOMAIN_0.cache.debug
`05-28-21 01:11:00 - (5542351) - [freedns] - Error 404 : Page not found`

**Environment**
OPNsense 21.1.6-amd64

**Resolution**
The URL in plugins/dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc is incorrect.
Line 567 contains the URL `https://sync.afraid.org/u/`
This should be `https://freedns.afraid.org/dynamic/update.php?`
For more info on the correct URL see: [https://freedns.afraid.org/dynamic/](https://freedns.afraid.org/dynamic/) (Login required)
When changed locally in /usr/local/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc this resolved the problem

**Comment by fichtner** [comment]
The change was submitted as a fix, but apparently it switched FreeDNS to v2 which requires an API token instead.

**Comment by sraspe** [comment]
I believe you allready needed your API key. It was allready described in the OPNsense Dynamic DNS manual: [https://docs.opnsense.org/manual/dynamic_dns.html](https://docs.opnsense.org/manual/dynamic_dns.html)

For anyone searching, you can find the API key from either the "direct URL" on the freedns site when you are logged in: [https://freedns.afraid.org/dynamic/](https://freedns.afraid.org/dynamic/): Everything after the `?` in the URL. Or you can download wget script or curl script fromt he same page which will contain the same URL.

**Comment by garudalicious** [comment]
I can second this fix works. Encountered the same issue and was already using API token. I edited phpDynDNS.inc as specified in the initial post and it's working as expected now.

**Comment by sraspe** [comment]
Hello @fichtner,
I think the v1 has been decommisioned by freedns. I just updated to 21.7.1 and cannot get the dynamicDNS to work without the fix.
By just replacing the URL in `/usr/local/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc` it works again. You can set the API key as the password (as in teh OPNsense documentaion) and it works flawlessly.
Could you please change the URL in `plugins/dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc` from `https://sync.afraid.org/u/` to `https://freedns.afraid.org/dynamic/update.php?`
when you get around to it? That should be all required to fix the integration.

Thanks!

**Comment by fichtner** [comment]
Whatever https://github.com/opnsense/plugins/pull/2263 was I don't know. If somebody complains about #2640 in the future FreeDNS will be removed from the plugin to avoid more bouncing.


Cheers,
Franco

**Comment by fichtner** [comment]
FWIW, vanilla paste from https://freedns.afraid.org/ front page:

```
Dynamic update demonstration example (v2 interface)
--
[~] $ curl https://sync.afraid.org/u/CyTXMbtq5cPnLjEg5vKHTPDE/Updated demo.freshdns.com from 107.170.238.X to 50.23.197.94
```

**Comment by sbellon** [comment]
I got bitten by this now when updating from 21.7.4 to 21.7.5_2. I had Dynamic DNS with freeDNS working with v2 and the API token. Now I wondered why it didn't work anymore.

Just using https://sync.afraid.org/u/TOKEN/ works just fine when enabling v2 in the freeDNS account. Not sure why this got reverted back to v1 ...

**Comment by fichtner** [comment]
Yeah as I said I will remove FreeDNS support if there is no clear compatibility between URLs and complaints keep coming in. This is clearly an issue with their service so I would encourage everyone to seek out their support to get the right url and get the other one fixed.

**Comment by sbellon** [comment]
If - for some reason - some people cannot update to freeDNS v2, then it could be an idea to provide both: "freeDNS v1" with the URL "https://freedns.afraid.org/dynamic/update.php?TOKENV1/" and a second "freeDNS v2" with the URL "https://sync.afraid.org/u/TOKENV2/". That way, everyone could the the freeDNS version he wants ...

**Comment by HelterBelter** [comment]
> If - for some reason - some people cannot update to freeDNS v2, then it could be an idea to provide both: "freeDNS v1" with the URL "https://freedns.afraid.org/dynamic/update.php?TOKENV1/" and a second "freeDNS v2" with the URL "https://sync.afraid.org/u/TOKENV2/". That way, everyone could use the freeDNS version he wants ...

Or have the user enter the complete URL of their choosing instead of hard-coding part of it.

**Comment by sbellon** [comment]
But then, the user would have to inform himself/herself - and additionally could make a typo when entering the URL. I don't see any problem with just providing the two alternative API versions available for freeDNS.

**Comment by HelterBelter** [comment]
If there are 2 versions offered the users have to inform themselves anyway.
And if they can copy/paste a token they can copy/paste a URL. It's offered right there on the FreeDNS website.

But either way is fine, both work for me. As long as it stops changing.

**Comment by fichtner** [comment]
The service to specify URLs is called "custom". Look, there is a huge issue at FreeDNS with interoperability and/or documentation and there is no point fixing this from here.

**Comment by sbellon** [comment]
I used a Ubiquiti Security Gateway for the last five years. That implemented FreeDNS v1. I had literally zero problems with that.

Then I switched to OPNsense two weeks ago (version 21.7.3). Suddenly FreeDNS v2 didn't work any more. I dug into that and found out that OPNsense obviously required FreeDNS v2 API to be used. I didn't know it up until that time. I enabled it in my FreeDNS account and voila, everything worked straight away.

Then I upgraded to OPNsense 21.7.5 just the other day and it stopped working suddenly again. I again, dug into the source of os-dyndns on github just to find out that the API was reverted back to v1.

I had zero issues with FreeDNS over the last five years. Neither with v1 nor the few days with v2.

I have not the history with OPNsense that - obviously! - you have, however I cannot agree that I had interoperability or documentation issues with FreeDNS the last five years.

But if there is a "Custom" to just put a URL ... that's fine with me as well.

**Comment by fichtner** [comment]
@sbellon If you want to imply user contributions to this plugin don't work for you that's fair enough, but I don't tend to agree. ;)

**Comment by sbellon** [comment]
I'm sorry, I didn't understand your last sentence.

Would you accept a PR with FreeDNS v1 and FreeDNS v2? Or are you just tired of FreeDNS and want to get rid of it, regardless.

I'm fine with both (since I realized that there is a Custom provider), just wondering whether it would be worth the hassle to prepare a patch...

**Comment by fichtner** [comment]
Going to https://freedns.afraid.org/ verifiably lists "v2" as the right URL. Generally speaking "v2" is also preferable to "v1". I don't mind flipping it once more to "v2" and be done with it. There's no need to do a PR because this is easily reverted.

And, again, people using this service really need to figure out what is going on with the URL trickery or how to handle their account over there. We can't keep supporting both things when something is clearly not working as intended (and even if it's just either people not knowing how to switch to "v2" or if there is such a thing as "v2" in the first place).

**Comment by sbellon** [comment]
Yes, I completely agree.

You could link to https://freedns.afraid.org/dynamic/v2/ in the description of the freeDNS plugin. I think this URL redirects you to the login page if you do not have an account, but if you have one, you get to see the explanation of v2, also with a link to another "tips" page and how to convert, etc. That's the reason I was a bit surprised to hear that you think there is no documentation. I think it's just that perhaps some users don't read it ... but ... isn't that true for everything? ;-)

**Comment by sraspe** [comment]
> I think this URL redirects you to the login page if you do not have an account, but if you have one, you get to see the explanation of v2, also with a link to another "tips" page and how to convert, etc. That's the reason I was a bit surprised to hear that you think there is no documentation. I think it's just that perhaps some users don't read it

I definitely agree that reading is difficult, but in this case I can't find it even after you pointed out where I should look!
The description in the tips is what URLs to use to update in v1 vs v2, but I don't see any way to set in my profile what interface version is active. I assumed they both are based on the descriptions and you can choose weather to update using v1 or v2. Also I don't see anything regarding this in the FAQ.
What am I missing?! I am logged-in (with a free account, I don't have premium).

**Comment by sbellon** [comment]
Sorry, I didn't want to come across offensive or dismissive.

I think, the afraid.org page changes as soon as you activated v2 for a host, so I am not sure myself how it looked when I activated it. However I still remember that on https://freedns.afraid.org/dynamic/ there was a link that lead to the v2. That link - at least for me right now - is https://freedns.afraid.org/dynamic/v2/ and on that page I see a table with my dynamic entries, can check/uncheck the entries and then select the action Enable/Disable for them which then gives the API v2 URL in the table.

Does it not look like this on your side? BTW: I do not have premium either.

**Comment by sraspe** [comment]
Thanks!
No offense taken, I was being sarcastic!

In my v2 interface the domain is set to "disabled" for me while the v1 stopped working.
I believe the check-mark and "Activating" it will only mark it to update all subdomains simultaneously:
`You can have multiple records updated with a single update by linking your records together (using the 'randomized update token' style). To link, clone your preferred update token by checking the box next to the update URL, then on the left hand side, check which records you want that identical token to be cloned to, then click Enable Dynamic DNS at the bottom of the page to apply.`
https://freedns.afraid.org/dynamic/v2/tips/

What really throws me off is that it worked until something happened in June when I originally created the bug report. At that time I didn't change my FreeDNS account, but the v1 interface stopped working. So I thought it was being decommissioned and created the bug report.
The only fix I could get working at that time was to update the URL in the phpDynDNS.inc. I checked a few times after that and the behavior was always the same (v1 broken, v2 working). However when I test now both v1 and v2 work again...

What I never expected is that for some people it is the other way around, v1 working, but v2 not.

Maybe time to follow Franco's advice and search for another service.

**Comment by sbellon** [comment]
Fair enough. However for me, freeDNS has always been reliable (five years using v1 and the last few weeks using v2), so I'm reluctant to switching services. I'm a big fan of "if it ain't broke, don't fix it". And from my perspective, it's not freeDNS that is broken.

**Comment by psychogun** [comment]
Same experience here. Dynamic updates was working - then it was not working - then I switched to v2 and got it working - now after an update, it does not work anymore.

My solution was to create a Custom service and these were the options I used:
* Enable: check
* Service type: Custom
* Interface to monitor: WAN
* Interface to send updates from: WAN
* Hostname: your.hostname.afraid.org
* CURL options: Force IPv4 resolving
* Password: Token
* Update URL: http://sync.afraid.org/u/
* Description: your.hostname.afraid.org

Hit Save and Force Update. 

You find your Token by going to https://freedns.afraid.org/dynamic/ 
Make sure to have enabled version 2 of the dynamic update interface:  https://freedns.afraid.org/dynamic/v2/

**Comment by sbellon** [comment]
Does this also detect success/failure of the dynamic DNS update?

**Comment by psychogun** [comment]
@sbellon yes, just confirmed with the default cron job "Dynamic DNS Update"
```
/usr/local/etc/rc.dyndns: Dynamic DNS: (Success) IP Address Updated Successfully!
```

**Comment by kintaroju** [comment]
so the question is for the next update will we have a thing to toggle between v1 and v2, or we back to just purely v1 again?

Also what was the reason why we went back using v1, does anyone know why?

**Comment by psychogun** [comment]
@kintaroju 

I think I will just use the Custom option instead of fiddling with v1 / v2 etc., as it seems to work just fine using 'Custom service with FreeDNS v2.

If v1 works with the Custom option as well, I'd opt in to remove FreeDNS and make a comment about it in the documentation.

**Comment by fichtner** [comment]
We will switch FreeDNS to v2 again, make a note and a pointer for this issue in release notes and leave it at that.


Cheers,
Franco

**Comment by psychogun** [comment]
@sbellon I think I spoke too soon. The logs says success, but the control panel in freedns does not update to the new IP. 

Anyways, I have updated the firewall, and now I am back on the FreeDNS option again =) Which works. 

Probably some custom information is missing in regards to the Custom service and FreeDNS. When I got the success message, I thought everything was OK- I was using the Token in the password field.
